### PR TITLE
feat(exex): subscribe to notifications with head using `ExExContext`

### LINKED ISSUE #11215: `ExExNotification::with_head` has bad API
### Describe the feature

The new `ExExNotifications::with_head` API takes `self` and as a result can't be used without deconstructing the `ExExContext`.

It will in fact do a partial move, but as a design pattern, people tend to keep the `ExExContext` as a field on their struct, so the partial move prevents further `&self` or `&mut self` method calls.

There are different possible ways to address this UX issue:
- Add an `Option` around notifications that we can take from and leave empty
- Add an abstraction over with/without head and have a method on the context to convert from without to with, changing the contents of that field without deconstructing

thanks @prestwich for reporting

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by shekhirin**
Closes https://github.com/paradigmxyz/reth/issues/11215

Similar to https://github.com/paradigmxyz/reth/pull/11495, but takes an approach of a newtype wrapper with a private enum, so that we don't expose the internal `Invalid` variant to the user.

### CODE REVIEW

File: crates/exex/exex/src/context.rs [hunk 1]
```diff
@@ -92,4 +93,29 @@ impl<Node: FullNodeComponents> ExExContext<Node> {
     pub fn task_executor(&self) -> &TaskExecutor {
         self.components.task_executor()
     }
+
+    /// Replaces notifications stream with a stream of notifications without a head.
+    ///
+    /// Consumes the type and returns a new instance of [`ExExContext`].
+    pub fn with_notifications_without_head(self) -> Self {
```
Comment (mattsse) [inline crates/exex/exex/src/context.rs:100]: isn't this the default?

Can we come up with a less confusing name? :)

File: crates/exex/exex/src/notifications.rs [hunk 2]
```diff
@@ -13,101 +13,194 @@ use std::{
...
-pub struct ExExNotifications<P, E> {
-    node_head: Head,
-    provider: P,
-    executor: E,
-    notifications: Receiver<ExExNotification>,
-    wal_handle: WalHandle,
-}
+/// A stream of [`ExExNotification`]s.
+#[derive(Debug)]
+pub struct ExExNotifications<P, E>(Option<ExExNotificationsInner<P, E>>);
```
Comment (mattsse) [inline crates/exex/exex/src/notifications.rs:18]: please convert this to a field

File: crates/exex/exex/src/notifications.rs [hunk 3]
```diff
@@ -13,101 +13,194 @@ use std::{
...
-        f.debug_struct("ExExNotifications")
-            .field("provider", &self.provider)
-            .field("executor", &self.executor)
-            .field("notifications", &self.notifications)
-            .finish()
-    }
+#[derive(Debug)]
+enum ExExNotificationsInner<P, E> {
+    WithoutHead(ExExNotificationsWithoutHead<P, E>),
+    WithHead(ExExNotificationsWithHead<P, E>),
```
Comment (mattsse) [inline crates/exex/exex/src/notifications.rs:23]: needs docs

File: crates/exex/exex/src/notifications.rs [hunk 4]
```diff
@@ -13,101 +13,194 @@ use std::{
...
 }
 
 impl<P, E> ExExNotifications<P, E> {
-    /// Creates a new instance of [`ExExNotifications`].
-    pub const fn new(
+    /// Creates a new stream of [`ExExNotifications`] without a head.
+    pub const fn new_without_head(
```
Comment (mattsse) [inline crates/exex/exex/src/notifications.rs:28]:
```suggestion
    pub const fn new(
```

File: crates/exex/exex/src/context.rs [hunk 5]
```diff
@@ -92,4 +93,29 @@ impl<Node: FullNodeComponents> ExExContext<Node> {
...
+        Self { notifications: self.notifications.without_head(), ..self }
+    }
+
+    /// Replaces notifications stream with a stream of notifications with a head using the
+    /// provided head.
+    ///
+    /// Consumes the type and returns a new instance of [`ExExContext`].
+    pub fn with_notifications_with_head(self, head: ExExHead) -> Self {
```
Comment (mattsse) [inline crates/exex/exex/src/context.rs:108]: this is a bit verbose, with_with_head

File: crates/exex/exex/src/notifications.rs (hunk 2 above)
Comment (mattsse) [inline crates/exex/exex/src/notifications.rs:18]: there's now no way to obtain the regular ExExNotificationsWithoutHead stream, right?

**Review by mattsse — CHANGES_REQUESTED** [review]
this makes sense, only have naming nits.

this makes the stream API worse for regular streams, but I think we can live with that.

perhalps we could combine this with parts of #11495 and replace `Option<ExexInner>` with the enum?

https://github.com/paradigmxyz/reth/pull/11495/files#diff-ffb8568607451c18b9e789d3626ae3379862cb33fc72b386e5419a79fa9dbe5fR29

File: crates/exex/exex/src/notifications.rs (hunk 3 above)
Comment (shekhirin) [reply crates/exex/exex/src/notifications.rs]: added

File: crates/exex/exex/src/notifications.rs (hunk 2 above)
Comment (shekhirin) [reply crates/exex/exex/src/notifications.rs]: done

File: crates/exex/exex/src/context.rs (hunk 1 above)
Comment (shekhirin) [reply crates/exex/exex/src/context.rs]: yep, removed these methods on the `ExExContext` and just left the mutating versions

File: crates/exex/exex/src/context.rs (hunk 5 above)
Comment (shekhirin) [reply crates/exex/exex/src/context.rs]: https://github.com/paradigmxyz/reth/pull/11500#discussion_r1790105958

File: crates/exex/exex/src/notifications.rs (hunk 2 above)
Comment (shekhirin) [reply crates/exex/exex/src/notifications.rs]: yes, no way, they fields and the `new` method are private

File: crates/exex/exex/src/notifications.rs [hunk 6]
```diff
@@ -13,101 +13,179 @@ use std::{
...
+            ExExNotificationsInner::Invalid => unreachable!(),
+            ExExNotificationsInner::WithoutHead(notifications) => notifications,
+            ExExNotificationsInner::WithHead(notifications) => ExExNotificationsWithoutHead::new(
+                notifications.node_head,
+                notifications.provider,
+                notifications.executor,
+                notifications.notifications,
+                notifications.wal_handle,
+            ),
+        });
```
Comment (mattsse) [inline crates/exex/exex/src/notifications.rs:88]: isn't this just self.set_without_head()?

File: crates/exex/exex/src/notifications.rs [hunk 7]
```diff
@@ -13,101 +13,179 @@ use std::{
...
+            ),
+            ExExNotificationsInner::WithHead(notifications) => ExExNotificationsWithHead::new(
+                notifications.node_head,
+                notifications.provider,
+                notifications.executor,
+                notifications.notifications,
+                notifications.wal_handle,
+                exex_head,
+            ),
+        });
```
Comment (mattsse) [inline crates/exex/exex/src/notifications.rs:130]: same pattern

File: crates/exex/exex/src/notifications.rs [hunk 8]
```diff
@@ -13,101 +13,179 @@ use std::{
...
-    /// has processed up to block 10, and is ready to process block 11.
-    pub fn with_head(self, head: ExExHead) -> ExExNotificationsWithHead<P, E> {
+    type Item = eyre::Result<ExExNotification>;
+
+    fn poll_next(
+        self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Option<Self::Item>> {
+        match &mut self.get_mut().inner {
+            ExExNotificationsInner::Invalid => unreachable!(),
```
Comment (mattsse) [inline crates/exex/exex/src/notifications.rs:147]: branches are checked in the order they appear, since this is unreachable, this should go last

**Review by mattsse — CHANGES_REQUESTED** [review]
nits

File: crates/exex/exex/src/notifications.rs (hunk 8 above)
Comment (shekhirin) [reply crates/exex/exex/src/notifications.rs]: good point, fixed

File: crates/exex/exex/src/notifications.rs (hunk 6 above)
Comment (shekhirin) [reply crates/exex/exex/src/notifications.rs]: yep, fixed

**Review by mattsse — APPROVED** [review]

**Pull request merged**